### PR TITLE
pacific: mgr/snap_schedule: backports

### PR DESCRIPTION
**Pull request by vshankar**
https://tracker.ceph.com/issues/53760
https://tracker.ceph.com/issues/55056

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by mchangir — COMMENTED** [review]
LGTM

**Comment by vshankar** [comment]
@yuriw I clubbed the remaining two CephFS PRs into this backport as they were kind of related. Please pick this up for pacific.

**Comment by vshankar** [comment]
These jenkins failures are super annoying :(

**Comment by vshankar** [comment]
jenkins retest this please

File: src/pybind/mgr/snap_schedule/fs/schedule.py [hunk 1]
```diff
@@ -233,6 +233,14 @@ def list_schedules(cls, path, db, fs, recursive):
...
             return [cls._from_db_row(row, fs) for row in c.fetchall()]
 
+    @classmethod
+    def list_all_schedules(cls,
+                           db: sqlite3.Connection,
+                           fs: str) -> List['Schedule']:
+        with db:
+            c = db.execute(cls.PROTO_GET_SCHEDULES + " path LIKE '%'")
+            return [cls._from_db_row(row, fs) for row in c.fetchall()]
+
```
Comment (mchangir) [inline src/pybind/mgr/snap_schedule/fs/schedule.py:243]: commit: **mgr/snap_schedule: restart old schedules**
the `List['Schedule']` type hint requires the `List` type imported from `typing`
please add that to the `import` section

Comment (vshankar) [reply src/pybind/mgr/snap_schedule/fs/schedule.py]: done

**Comment by vshankar** [comment]
@yuriw pls wait for jenkins test to pass

**Comment by vshankar** [comment]
Test run: https://pulpito.ceph.com/vshankar-2022-04-18_06:25:04-fs-wip-yuri4-testing-2022-04-15-1307-pacific-testing-default-smithi/

Failures are: 

```
"2022-04-18T06:51:11.867145+0000 mon.a (mon.0) 410 : cluster [WRN] Health check failed: Reduced data availability: 2 pgs peering (PG_AVAILABILITY)" in cluster log
```

Test cases run fine:
```
2022-04-18T07:01:52.624 INFO:tasks.cephfs_test_runner:Test concurrent snap creates in same file-system without db issues ... ok
2022-04-18T07:04:15.424 INFO:tasks.cephfs_test_runner:Test exisitence of multiple scheduled snapshots ... ok
2022-04-18T07:05:01.533 INFO:tasks.cephfs_test_runner:Test listing non-existing snap schedules failure ... ok
2022-04-18T07:05:43.496 INFO:tasks.cephfs_test_runner:Test listing snap schedules on a non-existing filesystem path failure ... ok
2022-04-18T07:07:14.981 INFO:tasks.cephfs_test_runner:Test existence of a scheduled snapshot ... ok
2022-04-18T07:07:59.454 INFO:tasks.cephfs_test_runner:Test listing snap schedules post removal of a schedule ... ok
2022-04-18T07:24:37.146 INFO:tasks.cephfs_test_runner:Test that snap schedule is resumed after mgr restart ... ok
2022-04-18T07:27:15.404 INFO:tasks.cephfs_test_runner:Test scheduled snapshots along with rentention policy ... ok
```

**Comment by vshankar** [comment]
commit 50281e0169bd08480cd767dca41daf02e2b73d1f is only for pacific.

@yuriw this is ready for testing.

File: qa/tasks/cephfs/test_snap_schedules.py [hunk 2]
```diff
@@ -438,9 +438,8 @@ def test_snap_schedule_with_mgr_restart(self):
         self.fs_snap_schedule_cmd('deactivate', path=testdir, snap_schedule='1M')
 
         new_stats = self.get_snap_stats(testdir)
-        self.assertTrue(new_stats['fs_count'] == new_stats['db_count'])
+        self.assertTrue(new_stats['fs_count'] == new_stats['db_count'] + old_stats['db_count'])
```
Comment (mchangir) [inline qa/tasks/cephfs/test_snap_schedules.py:441]: thanks for taking up this one @vshankar

Comment (vshankar) [reply qa/tasks/cephfs/test_snap_schedules.py:441]: Thank you for the initial backport. It was a tricky one.

**Comment by vshankar** [comment]
Reviewed-by: Venky Shankar <vshankar@redhat.com>

**Review by yuriw — APPROVED** [review]
Reviewed-by: Venky Shankar <vshankar@redhat.com>

**Pull request merged**